Frameworks/CoreGraphics/CGDataProvider.mm

```diff
@@ -16,130 +16,395 @@
 
 #import <StubReturn.h>
 #import <Starboard.h>
-#import <Foundation/NSData.h>
-#import <Foundation/NSString.h>
-#include "LoggingNative.h"
 #import <CoreGraphics/CGDataProvider.h>
 #import "CGDataProviderInternal.h"
+#import <CFCppBase.h>
 
-static const wchar_t* TAG = L"CGDataProvider";
+// Required so we can use CFHashBytes rather than trying our own hash algorithm
+#import <ForFoundationOnly.h>
 
-// TODO: Task 7188763-Reimplement CGDataProvider which does not inherit from NSData but rather uses it within itself.
-@implementation CGDataProvider : NSData
-- (void)dealloc {
-    if (releaseFunc != 0) {
-        releaseFunc(info, data, size);
+#import <vector>
+
+// Determines what amount of bytes to try and read at a time for indirect access data providers
+static constexpr size_t c_streamSize = 2048;
+
+#pragma region __CGDataProviderInternal
+namespace __CGDataProviderInternal {
+class __CGDataProviderInternal {
+    mutable std::unique_ptr<CFHashCode> m_hashValue;
+
+public:
+    virtual ~__CGDataProviderInternal() {
     }
-    filename = nil;
-    [super dealloc];
-}
+    virtual const void* GetData() const = 0;
+    virtual size_t GetSize() const = 0;
+    bool operator==(const __CGDataProviderInternal& rhs) const {
+        if (GetSize() != rhs.GetSize()) {
+            return false;
+        }
 
-- (NSString*)_fileName {
-    return filename;
-}
+        return memcmp(GetData(), rhs.GetData(), GetSize()) == 0;
+    }
 
-- (instancetype)init {
-    return [self initWithBytes:"" length:0];
-}
+    CFHashCode Hash() const {
+        if (!m_hashValue) {
+            m_hashValue.reset(new CFHashCode(CFHashBytes((UInt8*)GetData(), GetSize())));
+        }
 
-- (instancetype)initWithBytesNoCopy:(void*)bytes length:(NSUInteger)length freeWhenDone:(BOOL)freeWhenDone {
-    if (self = [super init]) {
-        _data.attach([[NSData alloc] initWithBytesNoCopy:bytes length:length freeWhenDone:freeWhenDone]);
+        return *m_hashValue;
     }
-    return self;
-}
+};
 
-- (const void*)bytes {
-    return [_data bytes];
-}
+class __UserDataProvider : public __CGDataProviderInternal {
+    mutable void* m_info;
+    const CGDataProviderCallbacks* m_callbacks;
+    mutable std::vector<unsigned char> m_data;
 
-- (NSUInteger)length {
-    return [_data length];
-}
+public:
+    __UserDataProvider(void* info, const CGDataProviderCallbacks* callbacks) : m_info(info), m_callbacks(callbacks) {
+    }
 
-@end
+    const void* GetData() const override {
+        if (m_data.empty()) {
+            size_t count = 0, delta = 0;
```

Comment:
there is a significant gain, which is reducing errors that can be accident;y introduced by a future author.
Also this is a standard no no situation within C++

---
In reply to: [93115701](https://github.com/Microsoft/WinObjC/pull/1579#discussion_r93115701) [](ancestors = 93115701,93115225,92931361)